csharp/src/Microsoft.ML.OnnxRuntime/FixedBufferOnnxValue.cs

```diff
@@ -43,6 +43,79 @@ public static FixedBufferOnnxValue CreateFromTensor<T>(Tensor<T> value)
+        /// but may also allow using custom types for Float16 and BFloat16 providing they have the same layout and size.
+        /// The resulting instance must be disposed of to release pinned memory and deallocate native OrtValue
+        /// See example below.
+        /// </summary>
+        /// <typeparam name="T">Blittable data type, compatible with supported types</typeparam>
+        /// <param name="memoryInfo">memoryInfo. For managed buffers simply use OrtMemoryInfo.DefaultInstance</param>
+        /// <param name="memory"></param>
+        /// <param name="elementType">TensorElementType</param>
+        /// <param name="shape">shape of the tensor to be created</param>
+        /// <param name="bytesSize">size of the allocation in bytes</param>
+        /// <returns>a disposable instance of FixedBufferOnnxValue</returns>
+        /// <example>
+        /// Here is an example of using a 3rd party library class for processing float16/bfloat16.
+        /// Currently, to pass tensor data and create a tensor one must copy data to Float16/BFloat16 structures
+        /// so DenseTensor can recognize it.
+        /// 
+        /// If you are using a library that has a class Half and it is blittable, that is its managed in memory representation
+        /// matches native one and its size is 16-bits, you can use the following conceptual example
+        /// to feed/fetch data for inference using Half array. This allows you to avoid copying data from your Half[] to Float16[]
+        ///
+        /// \code{.cs}
+        /// unsafe { Debug.Assert(sizeof(ushort) == sizeof(Half)); }
+        /// Half[] input = new Half[] { 5646, 12345 };
+        /// var input_shape = new long[] {input.Length};
+        /// Half[] output = new Half[40]; // Whatever the expected len/shape is must match
+        /// var output_shape = new long[] {output.Length};
+        /// 
+        /// var memInfo = OrtMemoryInfo.DefaultInstance; // CPU
+        ///
+        /// using(var fixedBufferInput = FixedBufferOnnxvalue.CreateFromMemory<Half>(memInfo,
+        ///                         input, TensorElementType.Float16, input_shape, input.Length * sizeof(ushort))
+        /// using(var fixedBufferOutput = FixedBufferOnnxvalue.CreateFromMemory<Half>(memInfo,
+        ///                               output, TensorElementType.Float16, output_shape, output.Length * sizeof(ushort))
+        /// {
+        ///    FixedBufferOnnxvalue[] inputValues = new FixedBufferOnnxvalue[]{fixedBufferInput};
+        ///    FixedBufferOnnxvalue[] outputValues = new FixedBufferOnnxvalue[]{fixedBufferOutput};
+        ///    session.Run(inputNames, inputValues, outputNames, outputValues);
+        ///   // Output is now in output[]
+        /// }
+        /// \endcode
+        /// </example>
+        public static FixedBufferOnnxValue CreateFromMemory<T>(OrtMemoryInfo memoryInfo, Memory<T> memory,
```

Comment:
Do we want to indicate that this applies only to tensors or is it too obvious ? I was thinking having an explicit name like CreateTensorFromMemory() ?

Reply:
I think it is mentioned in the comments

---
In reply to: [529142738](https://github.com/microsoft/onnxruntime/pull/5886#discussion_r529142738) [](ancestors = 529142738)